sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -1069,6 +1069,16 @@ object SQLConf {
       .checkValue(v => Set(1, 2).contains(v), "Valid versions are 1 and 2")
       .createWithDefault(2)
 
+  val STREAMING_JOIN_STATE_FORMAT_VERSION =
```

Comment:
I'm a little confused about why this needs to be a config. Isn't this just metadata written to the state store? Why would a user ever need to set this?

I have to go over the rest of the code with more care, since I'm really not familiar with it. But it looks like an insane amount of code just to add one single field to some state object...

Reply:
That is internal config and end users don't need to set this. This approach has been picked to mark the value of config into metadata and restrict changing the config once the query has been loaded from checkpoint. Please refer FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION /  STREAMING_JOIN_STATE_FORMAT_VERSION to see how they handled.

When we deal with modification of schema for state, unless we reach a consensus to let end users just discard checkpoint before continuing (that's totally destructive so we tend to avoid this), we should construct a way to deal with both schemas: that's why the concept of "versioning" comes out, and additional interfaces/implementations are introduced to handle these versions. So the actual code diff gets much bigger than what the patch is originally trying to fix.

Reply:
Referring previous patch just to add rationalization of bigger code diff here: please refer #21739 and compare diff of schemas of state row in StateManagerImplV1 and StateManagerImplV2 with lines of code change.

Reply:
I still don't get why this needs to be a config.

Shouldn't the version be recorded in the state store itself? As in, when you open it and read its data, there's a version number recorded there. (Or if there isn't, then it's the "legacy" version where the version was not written.)

Reply:
What you see "version" in state store denotes "batchId", which is different from semantic of "version" in here which denotes the difference of schema - I know that's confusing - not sure why "version" is taken instead of "batch" or "epoch" in state store. We already mixed up multiple terms to denote the same (or similar - epoch is not strictly same as batch when it's being used in continuous mode, as there's no micro-batch).

Reply:
I think I understand what's going on in the state store after looking at the code again.